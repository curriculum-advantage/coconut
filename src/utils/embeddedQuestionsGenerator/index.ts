import { prandoShuffle } from '../questionGeneration';

const getShuffledAnswerChoices = (
  numberGenerator,
  forceShuffledChoices,
  { distractors, correctAnswer },
) => {
  const answerOptions = distractors ? [...correctAnswer, ...distractors] : correctAnswer;
  return prandoShuffle(numberGenerator, answerOptions, forceShuffledChoices);
};

const getOrderedQuestions = ({
  numberGenerator,
  availableQuestions,
  requiredQuestionCount,
  shuffledQuestions,
}) => {
  const questions = shuffledQuestions
    ? prandoShuffle(numberGenerator, availableQuestions, false)
    : availableQuestions;

  return questions.slice(0, requiredQuestionCount);
};

const formattedStandardQuestions = (orderedQuestions, numberGenerator, forceShuffledChoices) => (
  orderedQuestions.map(question => {
    const answerChoices = getShuffledAnswerChoices(numberGenerator, forceShuffledChoices, question);
    return { ...question, answerChoices };
  })
);

const formattedGroupedQuestions = (orderedQuestions, numberGenerator, forceShuffledChoices) => (
  orderedQuestions.reduce((accumulator, section) => {
    const { common, group } = section;

    const sectionQuestions = group.map(question => {
      const answerChoices = getShuffledAnswerChoices(
        numberGenerator,
        forceShuffledChoices,
        question,
      );

      return { ...question, ...common, answerChoices };
    });

    return [...accumulator, ...sectionQuestions];
  }, [])
);

const getFormattedQuestions = ({
  numberGenerator,
  shuffledChoices,
  forceShuffledChoices,
  orderedQuestions,
  groupedQuestions,
}) => {
  let formattedQuestions = orderedQuestions;

  if (shuffledChoices) {
    const formatter = groupedQuestions ? formattedGroupedQuestions : formattedStandardQuestions;
    formattedQuestions = formatter(orderedQuestions, numberGenerator, forceShuffledChoices);
  }

  return formattedQuestions;
};

/**
 * Creates an embedded question generator (for static questions that are added to a JSON file).
 *
 * @param options Options object.
 * @param options.numberGenerator Pseudo random number generator created with Prando.
 * @param options.availableQuestions All possible questions that the activity can have.
 * @param options.requiredQuestionCount How many questions it takes to complete the activity.
 * @param options.shuffledQuestions Whether or not the questions should be shuffled.
 * @param options.shuffledChoices Whether or not the answer choices should be shuffled. If this
 * option is set to false, you should manually add answer choices (with the desired order) on to
 * each question in the JSON.
 * @param options.forceShuffledChoices Whether or not the answer choices should always be shuffled
 * until they do not match the order in which they were passed in. This can be useful in situations
 * where the correct answer is an order, assuming you don't want the user to have the answer
 * presented to them in the right order at any time. This option should be used strategically,
 * rather than habitually, since it can cause the correct answer to be predictable (e.g. answer
 * always on right side when shuffling 2-choice questions that are already ordered). See unit tests
 * for examples.
 * @param options.groupedQuestions Whether or not the questions are formatted into groups. See unit
 * tests for examples.
 *
 * @example
 *
 * embeddedQuestionsGenerator({
 *   numberGenerator: new Prando('someSeed'),
 *   availableQuestions: [{ "stem": "1 + 2", "correctAnswer": [3], "distractors": [4] }],
 * });
 */
const embeddedQuestionsGenerator = ({
  numberGenerator,
  availableQuestions,
  requiredQuestionCount = availableQuestions.length,
  shuffledQuestions = true,
  shuffledChoices = true,
  forceShuffledChoices = false,
  groupedQuestions = false,
}) => (
  {
    getQuestions: () => {
      // Shuffle if necessary, and return only the number of questions needed.
      const orderedQuestions = getOrderedQuestions({
        numberGenerator,
        availableQuestions,
        requiredQuestionCount,
        shuffledQuestions,
      });

      // Add answer choices on to each question.
      const formattedQuestions = getFormattedQuestions({
        numberGenerator,
        shuffledChoices,
        forceShuffledChoices,
        orderedQuestions,
        groupedQuestions,
      });

      return formattedQuestions;
    },
  }
);

export default embeddedQuestionsGenerator;
