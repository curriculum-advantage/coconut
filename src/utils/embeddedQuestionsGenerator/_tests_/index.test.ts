import Prando from 'prando';
import embeddedQuestionsGenerator from '..';

const seed = 'Test_Seed';

const standardQuestions = [
  { stem: '1 + 2', correctAnswer: [3], distractors: [1, 2, 4, 5] },
  { stem: '3 + 2', correctAnswer: [5], distractors: [1, 2, 3, 4] },
  { stem: '1 + 1', correctAnswer: [2], distractors: [1, 3, 4, 5] },
  { stem: '1 + 0', correctAnswer: [1], distractors: [2, 3, 4, 5] },
  { stem: '2 + 2', correctAnswer: [4], distractors: [1, 2, 3, 5] },
];

const twoAnswerChoiceQuestions = [
  { stem: '1 + 2', correctAnswer: [3], distractors: [1] },
  { stem: '3 + 2', correctAnswer: [5], distractors: [1] },
  { stem: '1 + 1', correctAnswer: [2], distractors: [1] },
  { stem: '1 + 0', correctAnswer: [1], distractors: [2] },
  { stem: '2 + 2', correctAnswer: [4], distractors: [1] },
];

const groupedQuestions = [
  {
    common: { paragraph: 'Story one text here.' },
    group: [
      {
        stem: 'Where does this story begin?',
        correctAnswer: ['Some place'],
        distractors: ['Some other place'],
      },
      {
        stem: 'When does this story take place?',
        correctAnswer: ['Some time'],
        distractors: ['Some other time'],
      },
    ],
  },
  {
    common: { paragraph: 'Story two text here.' },
    group: [
      {
        stem: 'Where does this story begin?',
        correctAnswer: ['Some place'],
        distractors: ['Some other place'],
      },
      {
        stem: 'When does this story take place?',
        correctAnswer: ['Some time'],
        distractors: ['Some other time'],
      },
    ],
  },
];

describe('embeddedQuestionsGenerator', () => {
  it('should return a shuffled version of the available questions being passed in, with shuffled answer choices (including all correct answers and distractors) added to each question', () => {
    const generator = embeddedQuestionsGenerator({
      numberGenerator: new Prando(seed),
      availableQuestions: standardQuestions,
    });
    expect(generator.getQuestions()).toMatchSnapshot();
  });

  it('should return only the number of questions specified by requiredQuestionCount', () => {
    const generator = embeddedQuestionsGenerator({
      numberGenerator: new Prando(seed),
      availableQuestions: standardQuestions,
      requiredQuestionCount: 1,
    });
    expect(generator.getQuestions()).toMatchSnapshot();
  });

  it('should keep questions in order if shuffledQuestions is set to false', () => {
    const generator = embeddedQuestionsGenerator({
      numberGenerator: new Prando(seed),
      availableQuestions: standardQuestions,
      shuffledQuestions: false,
    });
    expect(generator.getQuestions()).toMatchSnapshot();
  });

  it('should not add answer choices to the questions if shuffledChoices is set to false', () => {
    const generator = embeddedQuestionsGenerator({
      numberGenerator: new Prando(seed),
      availableQuestions: standardQuestions,
      shuffledChoices: false,
    });
    expect(generator.getQuestions()).toMatchSnapshot();
  });

  it('should allow shuffled answer choices to organically end up in the same order that they were passed in, when forceShuffledChoices is left at its default state of false', () => {
    const generator = embeddedQuestionsGenerator({
      numberGenerator: new Prando(seed),
      availableQuestions: twoAnswerChoiceQuestions,
    });
    expect(generator.getQuestions()).toMatchSnapshot();
  });

  it('should ensure that answer choices are shuffled in a way that they are never in the same order in which they were passed in, when forceShuffledChoices is set to true', () => {
    const generator = embeddedQuestionsGenerator({
      numberGenerator: new Prando(seed),
      availableQuestions: twoAnswerChoiceQuestions,
      forceShuffledChoices: true,
    });
    expect(generator.getQuestions()).toMatchSnapshot();
  });

  it('should support questions passed in as a grouped format, while keeping grouped questions together during the shuffling process, when groupedQuestions is set to true', () => {
    const generator = embeddedQuestionsGenerator({
      numberGenerator: new Prando(seed),
      availableQuestions: groupedQuestions,
      groupedQuestions: true,
    });
    expect(generator.getQuestions()).toMatchSnapshot();
  });
});
