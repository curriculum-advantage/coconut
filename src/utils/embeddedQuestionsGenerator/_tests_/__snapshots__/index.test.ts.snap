// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`embeddedQuestionsGenerator should allow shuffled answer choices to organically end up in the same order that they were passed in, when forceShuffledChoices is left at its default state of false 1`] = `
Array [
  Object {
    "answerChoices": Array [
      2,
      1,
    ],
    "correctAnswer": Array [
      2,
    ],
    "distractors": Array [
      1,
    ],
    "stem": "1 + 1",
  },
  Object {
    "answerChoices": Array [
      1,
      5,
    ],
    "correctAnswer": Array [
      5,
    ],
    "distractors": Array [
      1,
    ],
    "stem": "3 + 2",
  },
  Object {
    "answerChoices": Array [
      4,
      1,
    ],
    "correctAnswer": Array [
      4,
    ],
    "distractors": Array [
      1,
    ],
    "stem": "2 + 2",
  },
  Object {
    "answerChoices": Array [
      3,
      1,
    ],
    "correctAnswer": Array [
      3,
    ],
    "distractors": Array [
      1,
    ],
    "stem": "1 + 2",
  },
  Object {
    "answerChoices": Array [
      2,
      1,
    ],
    "correctAnswer": Array [
      1,
    ],
    "distractors": Array [
      2,
    ],
    "stem": "1 + 0",
  },
]
`;

exports[`embeddedQuestionsGenerator should ensure that answer choices are shuffled in a way that they are never in the same order in which they were passed in, when forceShuffledChoices is set to true 1`] = `
Array [
  Object {
    "answerChoices": Array [
      1,
      2,
    ],
    "correctAnswer": Array [
      2,
    ],
    "distractors": Array [
      1,
    ],
    "stem": "1 + 1",
  },
  Object {
    "answerChoices": Array [
      1,
      5,
    ],
    "correctAnswer": Array [
      5,
    ],
    "distractors": Array [
      1,
    ],
    "stem": "3 + 2",
  },
  Object {
    "answerChoices": Array [
      1,
      4,
    ],
    "correctAnswer": Array [
      4,
    ],
    "distractors": Array [
      1,
    ],
    "stem": "2 + 2",
  },
  Object {
    "answerChoices": Array [
      1,
      3,
    ],
    "correctAnswer": Array [
      3,
    ],
    "distractors": Array [
      1,
    ],
    "stem": "1 + 2",
  },
  Object {
    "answerChoices": Array [
      2,
      1,
    ],
    "correctAnswer": Array [
      1,
    ],
    "distractors": Array [
      2,
    ],
    "stem": "1 + 0",
  },
]
`;

exports[`embeddedQuestionsGenerator should keep questions in order if shuffledQuestions is set to false 1`] = `
Array [
  Object {
    "answerChoices": Array [
      2,
      1,
      5,
      3,
      4,
    ],
    "correctAnswer": Array [
      3,
    ],
    "distractors": Array [
      1,
      2,
      4,
      5,
    ],
    "stem": "1 + 2",
  },
  Object {
    "answerChoices": Array [
      1,
      2,
      3,
      4,
      5,
    ],
    "correctAnswer": Array [
      5,
    ],
    "distractors": Array [
      1,
      2,
      3,
      4,
    ],
    "stem": "3 + 2",
  },
  Object {
    "answerChoices": Array [
      2,
      1,
      4,
      5,
      3,
    ],
    "correctAnswer": Array [
      2,
    ],
    "distractors": Array [
      1,
      3,
      4,
      5,
    ],
    "stem": "1 + 1",
  },
  Object {
    "answerChoices": Array [
      1,
      3,
      4,
      5,
      2,
    ],
    "correctAnswer": Array [
      1,
    ],
    "distractors": Array [
      2,
      3,
      4,
      5,
    ],
    "stem": "1 + 0",
  },
  Object {
    "answerChoices": Array [
      3,
      2,
      1,
      4,
      5,
    ],
    "correctAnswer": Array [
      4,
    ],
    "distractors": Array [
      1,
      2,
      3,
      5,
    ],
    "stem": "2 + 2",
  },
]
`;

exports[`embeddedQuestionsGenerator should not add answer choices to the questions if shuffledChoices is set to false 1`] = `
Array [
  Object {
    "correctAnswer": Array [
      2,
    ],
    "distractors": Array [
      1,
      3,
      4,
      5,
    ],
    "stem": "1 + 1",
  },
  Object {
    "correctAnswer": Array [
      5,
    ],
    "distractors": Array [
      1,
      2,
      3,
      4,
    ],
    "stem": "3 + 2",
  },
  Object {
    "correctAnswer": Array [
      4,
    ],
    "distractors": Array [
      1,
      2,
      3,
      5,
    ],
    "stem": "2 + 2",
  },
  Object {
    "correctAnswer": Array [
      3,
    ],
    "distractors": Array [
      1,
      2,
      4,
      5,
    ],
    "stem": "1 + 2",
  },
  Object {
    "correctAnswer": Array [
      1,
    ],
    "distractors": Array [
      2,
      3,
      4,
      5,
    ],
    "stem": "1 + 0",
  },
]
`;

exports[`embeddedQuestionsGenerator should return a shuffled version of the available questions being passed in, with shuffled answer choices (including all correct answers and distractors) added to each question 1`] = `
Array [
  Object {
    "answerChoices": Array [
      1,
      3,
      4,
      5,
      2,
    ],
    "correctAnswer": Array [
      2,
    ],
    "distractors": Array [
      1,
      3,
      4,
      5,
    ],
    "stem": "1 + 1",
  },
  Object {
    "answerChoices": Array [
      5,
      1,
      3,
      4,
      2,
    ],
    "correctAnswer": Array [
      5,
    ],
    "distractors": Array [
      1,
      2,
      3,
      4,
    ],
    "stem": "3 + 2",
  },
  Object {
    "answerChoices": Array [
      4,
      2,
      3,
      5,
      1,
    ],
    "correctAnswer": Array [
      4,
    ],
    "distractors": Array [
      1,
      2,
      3,
      5,
    ],
    "stem": "2 + 2",
  },
  Object {
    "answerChoices": Array [
      4,
      2,
      1,
      3,
      5,
    ],
    "correctAnswer": Array [
      3,
    ],
    "distractors": Array [
      1,
      2,
      4,
      5,
    ],
    "stem": "1 + 2",
  },
  Object {
    "answerChoices": Array [
      1,
      3,
      5,
      2,
      4,
    ],
    "correctAnswer": Array [
      1,
    ],
    "distractors": Array [
      2,
      3,
      4,
      5,
    ],
    "stem": "1 + 0",
  },
]
`;

exports[`embeddedQuestionsGenerator should return only the number of questions specified by requiredQuestionCount 1`] = `
Array [
  Object {
    "answerChoices": Array [
      1,
      3,
      4,
      5,
      2,
    ],
    "correctAnswer": Array [
      2,
    ],
    "distractors": Array [
      1,
      3,
      4,
      5,
    ],
    "stem": "1 + 1",
  },
]
`;

exports[`embeddedQuestionsGenerator should support questions passed in as a grouped format, while keeping grouped questions together during the shuffling process, when groupedQuestions is set to true 1`] = `
Array [
  Object {
    "answerChoices": Array [
      "Some other place",
      "Some place",
    ],
    "correctAnswer": Array [
      "Some place",
    ],
    "distractors": Array [
      "Some other place",
    ],
    "paragraph": "Story two text here.",
    "stem": "Where does this story begin?",
  },
  Object {
    "answerChoices": Array [
      "Some time",
      "Some other time",
    ],
    "correctAnswer": Array [
      "Some time",
    ],
    "distractors": Array [
      "Some other time",
    ],
    "paragraph": "Story two text here.",
    "stem": "When does this story take place?",
  },
  Object {
    "answerChoices": Array [
      "Some place",
      "Some other place",
    ],
    "correctAnswer": Array [
      "Some place",
    ],
    "distractors": Array [
      "Some other place",
    ],
    "paragraph": "Story one text here.",
    "stem": "Where does this story begin?",
  },
  Object {
    "answerChoices": Array [
      "Some other time",
      "Some time",
    ],
    "correctAnswer": Array [
      "Some time",
    ],
    "distractors": Array [
      "Some other time",
    ],
    "paragraph": "Story one text here.",
    "stem": "When does this story take place?",
  },
]
`;
